postpone model import in `channels.auth` to avoid ImproperlyConfigured error

yedpodtrzitko opened the pull request:
Hi,
the import of `channels.auth` is triggering import of `AnonymousUser` model, which should not happen before ASGI app is ready. Doing so raises `ImproperlyConfigured` exception which might be tricky for newcomers to solve (see numerous Issues and a few MR trying to pinpoint this in docs, eg. #1681, #1691)

To make it work, some code needs to be mixed among imports (`django.setup()` or `get_asgi_application()`), which does not look nice.

Postponing the model import can solve the most common culprit of this error.

carltongibson commented on channels/auth.py:
```diff
@@ -24,6 +23,9 @@ def get_user(scope):
     Return the user model instance associated with the given scope.
     If no user is retrieved, return an instance of `AnonymousUser`.
     """
+    # postpone model import to avoid ImproperlyConfigured error before ASGI app is ready
```
`...before Django setup is complete`? Maybe.

yedpodtrzitko replied:
done

carltongibson reviewed (commented):
Hi @yedpodtrzitko — Thanks for this, yes, I think it's probably a good idea.

yedpodtrzitko commented:
I moved the docs changes from #1693 here, as it makes sense to have it merged atomically.

- since `channels.auth` can be imported early now, I removed the mention of `AuthMiddlewareStack` as a problematic code and moved its import above `get_asgi_application()`
- the code sample in `routing.rst` looks like it would likely lead to the `ImproperlyConfigured` error too, since it's calling `get_asgi_application()` very late, so I replaced it with the example from `deploying.rst` to make it more unified.
- fixed a few typos

carltongibson commented:
Ok, thanks. I just need to have a little play. 👍

yedpodtrzitko commented:
@carltongibson I assume you're super busy, but is there any chance merging this soon? This seems like the most common issue newcomers are hitting and it's spawning a lot of redundant Issues and MRs (eg. fresh #1707)...

carltongibson commented:
Hi @yedpodtrzitko — yes, I'm just getting some bandwidth to swing back here now. This is on my agenda for the next point release. (Thanks! 👍)

carltongibson reviewed (approved):
Hi @yedpodtrzitko — sorry for the slow uptake here. This looks great. Thanks!
